### PR TITLE
[FEATURE] Add SCC support to CEA-708 decoder

### LINKED ISSUE #1423: [BUG] SCC extraction from mov file results in an empty SCC file
CCExtractor version:0.94

- regression?{no}
- What platform did you use? {Mac 12.2.1}
- What were the used arguments? `ccextractor -in=mp4 -out=scc -nofc -dru /Volumes/Drive5/Whacked/WhackedOutVideos_short.mov -o /Volumes/Drive5/Whacked/WhackedOutVideos_short.scc'

# Video links
http://videopress.s3.amazonaws.com/samples/Whacked/WhackedOutVideos_short.mov

# Additional information
A .scc file is created, but it has no captions. Captions are displayed in the terminal window in SRT format while the job is running. Tried with multiple mov files. Terminal output here:

ccextractor -in=mp4 -out=scc -nofc -dru /Volumes/Drive5/Whacked/WhackedOutVideos_short.mov -o /Volumes/Drive5/Whacked/WhackedOutVideos_short.scc
CCExtractor 0.94, Carlos Fernandez Sanz, Volker Quetschke.
Teletext portions taken from Petr Kutalek's telxcc
--------------------------------------------------------------------------
Input: /Volumes/Drive5/Whacked/WhackedOutVideos_short.mov
[Extract: 1] [Stream mode: MP4]
[Program : Auto ] [Hauppage mode: No] [Use MythTV code: Auto]
[CEA-708: 63 decoders active]
[CEA-708: using charset "none" for all services]
[Timing mode: Auto] [Debug: No] [Buffer input: No]
[Use pic_order_cnt_lsb for H.264: No] [Print CC decoder traces: No]
[Target format: .scc] [Encoding: UTF-8] [Delay: 0] [Trim lines: No]
[Add font color data: No] [Add font typesetting: Yes]
[Convert case: No][Filter profanity: No] [Video-edit join: No]
[Extraction start time: not set (from start)]
[Extraction end time: not set (to end)]
[Live stream: No] [Clock frequency: 90000]
[Teletext page: Autodetect]
[Start credits text: None]
[Quantisation-mode: CCExtractor's internal function]

-----------------------------------------------------------------
Opening file: /Volumes/Drive5/Whacked/WhackedOutVideos_short.mov
Analyzing data with GPAC (MP4 library)
Opening '/Volumes/Drive5/Whacked/WhackedOutVideos_short.mov': ok
Track 1, type=soun subtype=in24
Track 2, type=soun subtype=in24
Track 3, type=soun subtype=in24
Track 4, type=soun subtype=in24
Track 5, type=tmcd subtype=tmcd
Track 6, type=vide subtype=avc1
Track 7, type=clcp subtype=c708
MP4: found 7 tracks: 1 avc and 1 cc
Processing track 1, type=soun subtype=in24
Processing track 2, type=soun subtype=in24
Processing track 3, type=soun subtype=in24
Processing track 4, type=soun subtype=in24
Processing track 5, type=tmcd subtype=tmcd
Processing track 6, type=vide subtype=avc1
Processing track 7, type=clcp subtype=c708
00:00:02:501 --> 00:00:02:501
00:00:02:502 --> 00:00:03:636
Or is this synchronized
00:00:03:637 --> 00:00:05:971
Or is this synchronized
stick poking?
00:00:05:972 --> 00:00:07:173
Or is this synchronized
stick poking?
Actually, those are the rods
00:00:07:174 --> 00:00:08:841
stick poking?
Actually, those are the rods
they use to locate lost people
00:00:08:842 --> 00:00:10:443
Actually, those are the rods
they use to locate lost people
or objects in up to 12 feet
00:00:10:444 --> 00:00:12:478
they use to locate lost people
or objects in up to 12 feet
of powder.
00:00:12:479 --> 00:00:13:512
or objects in up to 12 feet
of powder.
In this case, they find this
00:00:13:513 --> 00:00:15:981
of powder.
In this case, they find this
dude's video camera.
00:00:15:982 --> 00:00:17:350
In this case, they find this
dude's video camera.
And the swear blizzard
00:00:17:351 --> 00:00:18:784
dude's video camera.
And the swear blizzard
starts again.
00:00:18:785 --> 00:00:22:755
And the swear blizzard
starts again.
(bleeping)
00:00:22:756 --> 00:00:24:623
starts again.
(bleeping)
Two-thirds of the world
00:00:24:624 --> 00:00:26:425
(bleeping)
Two-thirds of the world
is covered by water
00:00:28:661 --> 00:00:28:661
Two-thirds of the world
is covered by water
and the rest is covered by us.
00:00:28:662 --> 00:00:30:029
is covered by water
and the rest is covered by us.
Time to gather up all the goofy
100%  |  00:29
Closing media: ok
Found 1 AVC track(s). Found 1 CC track(s).

Total frames time:	  00:00:00:000  (0 frames at 29.97fps)

Min PTS:				00:00:00:000
Max PTS:				00:00:29:996
Length:				 00:00:29:996
Done, processing time = 0 seconds
00:00:30:030 --> 00:00:30:029
and the rest is covered by us.
Time to gather up all the goofy
goings-on from around the globe

No captions were found in input.
Issues? Open a ticket here

### LINKED ISSUE #1426: [PROPOSAL] Add SCC support to CEA-708 decoder
Add support for SCC format to CEA-708 decoder. 
Currently, only SRT, SAMI and Transcript formats are supported, https://github.com/CCExtractor/ccextractor/blob/master/src/rust/src/decoder/tv_screen.rs#L126-L134

SCC format details :- http://www.theneitherworld.com/mcpoodle/SCC_TOOLS/DOCS/SCC_FORMAT.HTML

#1423

### PR DESCRIPTION
**Pull request by IshanGrover2004**
**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I have considered, and confirmed that this submission will be valuable to others.
- [x] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [x] I give this submission freely, and claim no ownership to its content.
- [x] **I have mentioned this change in the [changelog](https://github.com/CCExtractor/ccextractor/blob/master/docs/CHANGES.TXT).**

**My familiarity with the project is as follows (check one):**

- [ ] I have never used CCExtractor.
- [x] I have used CCExtractor just a couple of times.
- [ ] I absolutely love CCExtractor, but have not contributed previously.
- [ ] I am an active contributor to CCExtractor.
---
I have tested some of the videos and it is working.
Closes #1426 (and in order to fix #1423)

### CODE REVIEW

**Comment by PunitLodha** [comment]
@IshanGrover2004 The SCC captions generated do not work well for the fixing the original issue #1423. Did you try fixing captions for the video provided there? https://github.com/CCExtractor/ccextractor/issues/1423#issuecomment-1073240342

**Comment by IshanGrover2004** [comment]
> The SCC captions generated do not work well for the fixing the original issue #1423. Did you try fixing captions for the video provided there? [#1423 (comment)](https://github.com/CCExtractor/ccextractor/issues/1423#issuecomment-1073240342)

Yeah, I tested that & that is because the captions are currently get overlapping in it & I did push the solution for overlapping 
I'll try to solve this by looking the correct SCC caption and my generated one & get back to you

**Comment by IshanGrover2004** [comment]
Hey @PunitLodha I did test and do some research and here is what i got
Captions is perfectly made, its just because of formatting issues 
And here is the [diff](https://www.diffchecker.com/m83uljVq/) of correct scc and scc generated by my code. Do have a look once. Do you know how can i decide using code when i need to place things like `94f2 94f2` to move cursor to row 15, column 4 ?

**Comment by PunitLodha** [comment]
> And here is the [diff](https://www.diffchecker.com/m83uljVq/) of correct scc and scc generated by my code.

Where did you get the correct SCC from? According to the original issue, this one seems to be correct, https://github.com/CCExtractor/ccextractor/issues/1426#issuecomment-1079731749

**Comment by IshanGrover2004** [comment]
> > And here is the [diff](https://www.diffchecker.com/m83uljVq/) of correct scc and scc generated by my code.
> 
> Where did you get the correct SCC from? According to the original issue, this one seems to be correct, [#1426 (comment)](https://github.com/CCExtractor/ccextractor/issues/1426#issuecomment-1079731749)

I don't remember the tool name but i have run it & its running correctly

File: src/rust/src/decoder/tv_screen.rs [hunk 1]
```diff
@@ -358,6 +360,100 @@ impl dtvcc_tv_screen {
...
+                    bytes_written += 1;
+                }
+                // add 0x80 padding and form byte pair if the last byte pair is not form
+                if bytes_written % 2 == 1 {
+                    buf.push_str("80 ");
+                } else {
+                    buf.push(' ');
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (IshanGrover2004) [reply src/rust/src/decoder/tv_screen.rs:448]: There are different codes for different positions in SCC.
I am also figuring out what code can I fit in there. In conclusion, There are some codes I have seen but I need to compare the lines or characters came in 1 row and do things accordingly.

Can I combine `self.chars` to the `String` kinda thing for experimental purpose bcs I tried but it is just giving me "0000...0"

File: src/rust/src/decoder/tv_screen.rs [hunk 2]
```diff
@@ -358,6 +360,140 @@ impl dtvcc_tv_screen {
...
+                    ones += 1;
+                }
+            }
+            if ones % 2 == 0 {
+                0b10000000 | value
+            } else {
+                value
+            }
+        }
+        fn add_needed_scc_labels(buf: &mut String, subtitle_count: usize, count: usize) {
```
Comment (PunitLodha) [inline src/rust/src/decoder/tv_screen.rs:384]:
```suggestion
        fn add_needed_scc_labels(buf: &mut String, total_subtitle_count: usize, subtitles_displayed: usize) {
```
Could you please add some documentation about this function, particularly mentioning in which cases it choose row 15 vs row 14 vs row 13

**Review by PunitLodha — APPROVED** [review]
Great work🎉

File: src/rust/src/decoder/tv_screen.rs (hunk 2 above)
Comment (IshanGrover2004) [reply src/rust/src/decoder/tv_screen.rs:384]: Yeah sure

Comment (IshanGrover2004) [reply src/rust/src/decoder/tv_screen.rs]: Done @PunitLodha

**Comment by IshanGrover2004** [comment]
Anything else left @PunitLodha ?

**Pull request merged**